T16940: add location label command and event

ramcq opened the pull request:
Introduce the event "eb0302d8-62e7-274b-365f-cd4e59103983" which is recorded at system startup and whenever location.conf is modified. The command eos-label-location is provided to prompt the administrator for the fields (such as facility, city and state) to store in this file. The event payload is a dictionary of string keys to the values provided in the location.conf file. The intention is to allow an operator to provide an optional human-readable label for the location of the system, which can be used when preparing reports or visualisations of the metrics data.

https://phabricator.endlessm.com/T16940

pwithnall commented on tools/Makefile.am.inc:
```diff
@@ -30,3 +30,19 @@ eos_network_id_CFLAGS = \
+bin_SCRIPTS = eos-label-location
+
+do_edit = sed \
+	-e "s|\@sysconfdir\@|$(sysconfdir)|"
```
Nitpick: I think you want the `g` flag at the end of that. The convention I’ve seen elsewhere for quoting the `@`s is: `-e 's|[@]sysconfdir[@]|$(sysconfdir)|g'`

pwithnall commented on tools/Makefile.am.inc:
```diff
@@ -30,3 +30,19 @@ eos_network_id_CFLAGS = \
+	-e "s|\@sysconfdir\@|$(sysconfdir)|"
+
+eos-label-location: tools/eos-label-location.in Makefile
+	$(AM_V_GEN) $(do_edit) $< >$@
```
Nitpick: Does `sed` do atomic overwrites, or does this rule need to write out to `$@.tmp && mv $@{.tmp,}`?

ramcq replied:
sed -i is atomic, but this was not, as relying on shell redirect due to different path - fixed.

pwithnall commented on tools/Makefile.am.inc:
```diff
@@ -30,3 +30,19 @@ eos_network_id_CFLAGS = \
 
 eos_network_id_LDADD = $(EOS_INSTRUMENTATION_LIBS)
+
+bin_SCRIPTS = eos-label-location
```
I think this needs a `nodist_` prefix.

pwithnall commented on tools/Makefile.am.inc:
```diff
@@ -30,3 +30,19 @@ eos_network_id_CFLAGS = \
+do_edit = sed \
+	-e "s|\@sysconfdir\@|$(sysconfdir)|"
+
+eos-label-location: tools/eos-label-location.in Makefile
```
Why not put the build output in the `tools/` directory (in `builddir`) as well? It’s a bit confusing for the source and build product to be in different directories.

ramcq replied:
Consistency with other stuff.

pwithnall replied:
⇒ Follow up PR! :-)

pwithnall commented on tools/eos-label-location.in:
```diff
@@ -0,0 +1,96 @@
+#
+# eos-label-location: edit the location label for metrics instrumentation daemon
+#
+# Copyright (C) 2017 Endless Mobile, Inc.
```
s/(C)/©/, for which you will also need to add the following as the second line of the file:
```
# -*- coding: utf-8 -*-
```

pwithnall commented on tools/eos-label-location.in:
```diff
@@ -0,0 +1,96 @@
+
+import argparse
+import configparser
+import os
```
How does this file fare against `pep8` and `python3-pylint`?

ramcq replied:
pep'd and flaked!

pwithnall commented on tools/eos-label-location.in:
```diff
@@ -0,0 +1,96 @@
+
+class LabelLocation(object):
+    CONFIG_DIR = '@sysconfdir@/metrics/'
+    CONFIG_FILE = CONFIG_DIR + 'location.conf'
```
Nitpick: Technically should be `os.path.join`.

ramcq replied:
abandoned the directory special casing so the full path is just hardcoded as the default to -f

pwithnall commented on tools/eos-label-location.in:
```diff
@@ -0,0 +1,96 @@
+    LABEL_SECTION = 'Label'
+    LABEL_KEYS = ['facility', 'city', 'state']
+
+    def __init__(self, args):
```
Why is `args` provided as a `dict` rather than as separate keyword arguments?

ramcq replied:
outright laziness, fixed

pwithnall commented on tools/eos-label-location.in:
```diff
@@ -0,0 +1,96 @@
+
+    def write(self):
+        if self.CONFIG_DIR:
+            os.makedirs(self.CONFIG_DIR, exist_ok=True)
```
Do we want to explicitly set the permissions on these directories?

ramcq replied:
abandoned this because it was for development, but I later added -f - there is no case this script should be used on a system that won't have an /etc/metrics (which has special permissions for the event recorder)

pwithnall commented on tools/eos-label-location.in:
```diff
@@ -0,0 +1,96 @@
+
+    def set(self):
+        for k in self.LABEL_KEYS:
+            old = ''
```
Nitpick: Would be a bit clearer to move this `old = ''` assignment to the `except` block below. Then you could drop the `pass`.

pwithnall commented on tools/eos-label-location.in:
```diff
@@ -0,0 +1,96 @@
+            the location of this computer, which will be sent to the Endless \
+            OS metrics system.')
+    parser.add_argument('-f', '--file', dest='file',
+                        help='Select config file (default: {})'.format(LabelLocation.CONFIG_FILE))
```
Would it help things here to use `type=argparse.FileType('rw')` for `--file`?

ramcq replied:
it turned out not to be so helpful, because we wish to set a default which it's OK to not exist at runtime and don't wish to create it as side-effect of executing the command with a different -f - but I've used default to good effect now at least

pwithnall commented on src/Makefile.am.inc:
```diff
@@ -40,6 +40,7 @@ eos_metrics_instrumentation_SOURCES = \
 eos_metrics_instrumentation_CPPFLAGS = \
 	$(EOS_INSTRUMENTATION_CFLAGS) \
 	-DINSTRUMENTATION_CACHE_DIR="\"$(instrumentationcachedir)\"" \
+	-DSYS_CONF_DIR="\"$(sysconfdir)\"" \
```
Nitpick: `-DSYSCONFDIR` would be a bit more greppable (`git grep -i sysconfdir`).

pwithnall commented on src/eos-metrics-instrumentation.c:
```diff
@@ -805,6 +815,90 @@ login_dbus_proxy_new (void)
+      {
+        /* this file’s existence is optional, so not found is not an error */
+        if (err->domain == G_KEY_FILE_ERROR &&
+            err->code == G_KEY_FILE_ERROR_NOT_FOUND)
```
`g_error_matches (err, G_KEY_FILE_ERROR, G_KEY_FILE_ERROR_NOT_FOUND)`

pwithnall commented on src/eos-metrics-instrumentation.c:
```diff
@@ -805,6 +815,90 @@ login_dbus_proxy_new (void)
+      return G_SOURCE_REMOVE;
+
+    g_autoptr (GString) label = g_string_new ("");
+    g_auto (GVariantBuilder) builder;
```
All `g_auto` variables *must* be initialised at declaration time, in case the function ends before they are assigned to, then we end up freeing garbage.

ramcq replied:
literally initialised on the next line, but OK whatever :P

pwithnall commented on src/eos-metrics-instrumentation.c:
```diff
@@ -805,6 +815,90 @@ login_dbus_proxy_new (void)
+    g_variant_builder_init (&builder, G_VARIANT_TYPE_ARRAY);
+    for (GStrv cur = keys; *cur != NULL; cur++)
+      {
+        gchar *key = *cur;
```
`const gchar *key`

pwithnall commented on src/eos-metrics-instrumentation.c:
```diff
@@ -805,6 +815,90 @@ login_dbus_proxy_new (void)
+    for (GStrv cur = keys; *cur != NULL; cur++)
+      {
+        gchar *key = *cur;
+        gchar *val = g_key_file_get_string (kf, LOCATION_LABEL_GROUP, key, NULL);
```
`g_autofree gchar *val`

ramcq replied:
good catch, was imagining it returning const pointers into the keyfile or similar

pwithnall commented on src/eos-metrics-instrumentation.c:
```diff
@@ -805,6 +815,90 @@ login_dbus_proxy_new (void)
+        if (val == NULL)
+          continue;
+
+        if (cur > keys)
```
Eek, numeric comparisons of pointers. `cur != keys` would be less scary.

pwithnall commented on src/eos-metrics-instrumentation.c:
```diff
@@ -805,6 +815,90 @@ login_dbus_proxy_new (void)
+        g_string_append_printf (label, "\"%s\" = \"%s\"", key, val);
+      }
+
+    g_message ("recording location label: %s", label->str);
```
Nitpick: s/recording/Recording/?

ramcq replied:
inconsistent with the other message we just merged :P

pwithnall replied:
:+1:

pwithnall commented on src/eos-metrics-instrumentation.c:
```diff
@@ -805,6 +815,89 @@ login_dbus_proxy_new (void)
+      return G_SOURCE_REMOVE;
+
+    g_autoptr (GString) label = g_string_new ("");
+    g_auto (GVariantBuilder) builder = { 0, };
```
`G_VARIANT_BUILDER_INIT` exists for this purpose, although it requires GLib 2.50 (I don’t know what your current dependency is).

pwithnall replied:
Looks good to merge once this is fixed.

ramcq replied:
Fixed.

cosimoc commented on src/eos-metrics-instrumentation.c:
```diff
@@ -805,6 +815,89 @@ login_dbus_proxy_new (void)
+
+    if (err != NULL)
+      {
+        g_warning ("couldn't set up file monitor for" LOCATION_CONF_FILE ": %s",
```
This will print with a missing space after `for` AFAICS

ramcq replied:
Fixed!

ramcq commented:
Actually @pwithnall you were right about the capitalisation of debug messages, so I went ahead and changed the other one. I also did the build changes and found a -Wpointer-signed nit which I added in the last PR.

pwithnall commented on .gitignore:
```diff
@@ -1,10 +1,10 @@
-/eos-label-location
-/eos-metrics-instrumentation
-/eos-network-id
+/src/eos-metrics-instrumentation
```
I still think you should use [`git.mk`](https://github.com/behdad/git.mk) in a follow-up PR or commit. It’s really really awesome.

ramcq replied:
done

pwithnall commented on tools/Makefile.am.inc:
```diff
@@ -16,33 +16,33 @@
 	$(AM_V_GEN) $(do_edit) $< >$@.tmp && mv $@{.tmp,}
 
-EXTRA_DIST = \
+EXTRA_DIST += \
```
Howcome this is `+=` now?

ramcq replied:
as these are included, the = was clobbering other values

pwithnall commented on tools/Makefile.am.inc:
```diff
@@ -16,33 +16,33 @@
 
-CLEANFILES = \
-	eos-label-location \
+CLEANFILES += \
```
Howcome this is `+=` now?

ramcq replied:
same
